src/ipa_pc/mod.rs

```diff
@@ -1204,146 +1228,225 @@ impl<G: AffineCurve, D: Digest> PolynomialCommitment<G::ScalarField> for InnerPr
+
+        let mut cur_challenge = G::ScalarField::one();
+
+        for (label, (_point_label, point)) in query_set.iter() {
+            let labeled_polynomial =
+                *poly_map.get(label).ok_or(Error::MissingPolynomial {
+                    label: label.to_string(),
+                })?;
+
+            let labeled_commitment =
+                *commitment_map.get(label).ok_or(Error::MissingCommitment {
+                    label: label.to_string(),
+                })?;
+
+            let labeled_randomness =
+                *rands_map.get(label).ok_or(Error::MissingRandomness {
+                    label: label.to_string(),
+                })?;
+
+            // (X - x_i)
+            let x_polynomial = Polynomial::from_coefficients_slice(&[
+                (G::ScalarField::zero() - point),
+                G::ScalarField::one()
+            ]);
+
+            // z_i(X) = z(X) / (X - x_i)
+            let z_i_polynomial = &z_polynomial / &x_polynomial;
+
+            // z_i(x)
+            let z_i_value = z_i_polynomial.evaluate(x_point);
+
+            // LC(p_1(X),p_2(X),...,p_m(X)) = SUM ( lamda^i * z_i(x) * p_i(X) )
+            lc_polynomial +=
+                (cur_challenge * z_i_value, labeled_polynomial.polynomial());
+
+            // LC(C) = SUM ( lambda^i * z_i(x) * C_i )
+            for (index, &comm_dash_segment) in labeled_commitment.commitment().comm.iter().enumerate() {
+                if index == lc_comm.len() {
+                    lc_comm.push(G::zero());
+                }
+                lc_comm[index] = (lc_comm[index].into_projective()
+                    + &(comm_dash_segment.into_projective()
+                    .mul(&z_i_value)
+                    .mul(&cur_challenge)
+                )
+                ).into_affine();
+            }
+
+            if has_hiding {
+                for (index, &rand_segment) in labeled_randomness.randomness().rand.iter().enumerate() {
+                    if index == lc_randomness.rand.len() {
+                        lc_randomness.rand.push(G::ScalarField::zero());
+                    }
+                    lc_randomness.rand[index] += rand_segment * cur_challenge * z_i_value;
+                }
+            }
+
+            // lambda^i
+            cur_challenge = cur_challenge * &lambda;
         }
 
-        // h(X) polynomial added to the set of polynomials for multi-poly single-point batching
-        let mut labeled_polynomials = labeled_polynomials;
+        // z(x)
+        let z_x_value = z_polynomial.evaluate(x_point);
+
+        // LC(p_1(X),p_2(X),...,p_m(X),h(X)) = SUM ( lamda^i * z_i(x) * p_i(X) ) - z(x) * h(X)
+        lc_polynomial += (-z_x_value, &h_polynomial);
+
+        // LC(C) = SUM ( lambda^i * z_i(x) * C_i ) - z(x) * C
+        lc_comm = lc_comm.iter().enumerate().map(|(index, &lc_comm_segment)| {
+            let h_comm_segment = if h_comm.len() > index {
+                h_comm[index]
+            } else {
+                G::zero()
```

Comment:
You could've also zipped two iterators on lc_comm and h_comm

Reply:
No, I cannot zip them, because h_comm and lc_comm can have different sizes. When we zipping 2 vectors of different sizes, the loop is truncated to the least one

Reply:
But you are summing G::zero() in case one is bigger than the other, so truncating would result in the same thing.

Reply:
Mmm, no, after that I am collecting the result back to the lc_comm, so if h_comm was smaller then lc_comm, lc_comm will be truncated to this size, and last elements will be lost